lib/BackgroundJob/ContextChat/SubmitContentJob.php

```diff
@@ -47,76 +55,112 @@ protected function run($argument): void {
-		$messageIds = array_filter($messageIds, fn (int $id): bool => $id >= $job->getNextMessageId());
+		$messages = $this->messageMapper->findByIds($account->getUserId(), $messageIds, 'asc', 'id');
 
-		if (count($messageIds) === 0) {
+		if (empty($messages)) {
+			try {
+				$this->jobsService->delete($task->getId());
+			} catch (MultipleObjectsReturnedException|Exception $e) {
+				$this->logger->warning('Exception occurred when trying to delete task', ['exception' => $e]);
+			}
 			return;
 		}
 
-		$messages = $this->messageMapper->findByIds($job->getUserId(), $messageIds, 'asc');
-		// Ensure messages are sorted by ID
-		usort($messages, static fn (Message $a, Message $b): int => $a->getId() <=> $b->getId());
 
-		$startTime = time() - ContextChatProvider::CONTEXT_CHAT_MESSAGE_MAX_AGE;
-		$nextMessage = reset($messages);
 		$client = $this->clientFactory->getClient($account);
 		$items = [];
 
 		try {
-			while (($nextMessage !== false) && (count($items) < ContextChatProvider::CONTEXT_CHAT_IMPORT_MAX_ITEMS)) {
-				// Skip older messages
-				if ($nextMessage->getSentAt() < $startTime) {
-					$nextMessage = next($messages);
+			$startTime = $this->time->getTime();
+			foreach ($messages as $message) {
+				if ($this->time->getTime() - $startTime > ContextChatProvider::CONTEXT_CHAT_MESSAGE_MAX_AGE) {
+					break;
+				}
+				try {
+					$imapMessage = $this->mailManager->getImapMessage($client, $account, $mailbox, $message->getUid(), true);
+				} catch (ServiceException $e) {
+					// couldn't load message, let's skip it. Retrying would be too costly
+					continue;
+				} catch (SmimeDecryptException $e) {
+					// encryption problem, skip this message
 					continue;
 				}
 
-				$imapMessage = $this->mailManager->getImapMessage($client, $account, $mailbox, $nextMessage->getUid(), true);
-
 				// Skip encrypted messages
 				if ($imapMessage->isEncrypted()) {
-					$nextMessage = next($messages);
 					continue;
 				}
 
 				$fullMessage = $imapMessage->getFullMessage($imapMessage->getUid(), true);
 
 				$items[] = new ContentItem(
-					(string)$nextMessage->getId(),
+					$mailbox->getId() . ':' . $message->getId(),
 					$this->contextChatProvider->getId(),
 					$imapMessage->getSubject(),
 					$fullMessage['body'] ?? '',
 					'E-Mail',
 					$imapMessage->getSentDate(),
-					[$job->getUserId()],
+					[$account->getUserId()],
 				);
-
-				$nextMessage = next($messages);
 			}
-		} finally{
-			$client->disconnect();
+		} finally {
+			try {
+				$client->close();
+			} catch (\Horde_Imap_Client_Exception $e) {
+				// pass
+			}
 		}
 
 		if (count($items) > 0) {
 			$this->contentManager->submitContent($this->contextChatProvider->getAppId(), $items);
 		}
 
-		if ($nextMessage !== false) {
-			$this->jobsService->updateOrCreate($job->getUserId(), $job->getAccountId(), $job->getMailboxId(), $nextMessage->getId());
+		try {
+			$this->jobsService->updateOrCreate($task->getMailboxId(), $message?->getId() ?? $messageIds[0]);
```

Comment:
Is there any situation where `$message` would not have an ID? And even then, why should we loop back to the start of the list?

Reply:
This is about `$message` being null and mainly to appease static analysis, I think. Feel free to improve this.